afpacket integration with route-rs

moosingin3space opened the pull request.

ErrHu committed:
FromPcap/ToPcap: add support to read/write pcap

Add the ability to read/write packets from pcap.
FromPcap offers two mode: Dump and Realtime
Dump - Pass next packets whenever polled
Realtime - Pass packets once their timestamp has passed

pcap uses libc::timeval, the definition of which changes
between architectures. So target_os is added to accomadate
the differences. This makes the pcap module os specific.

This was tested with a pcap of a ping to 8.8.8.8

ddowl committed:
runtime: Create log processors

CollinValley committed:
Packets: Create `packet` marker trait

We need a generic trait with which to refer to all underlying packet
types that are supported by packets. This is useful for some generic
wrapper structs in our router examples.

moosingin3space committed:
afpacket: add a split() function

Inspired by tokio::split, I have implemented splitting specifically for
afpacket. It exposes enough low-level functionality to implement the
route-rs-runtime links.

CollinValley committed:
NOCOMMIT: Disable graphgen tests

This is being done to facilitate changing the classifier schema to allow
dropping packets inside the classifier.

CollinValley committed:
Runtime: ClassifyLink: Allow dropping packets

In the past, ClassifyLink was unable to drop packets. This means that if
classify is being used as a complex filter (say sorting valid IP
traffic), it was unable to drop packets. It would have to send packets
to be dropped out a specific port, which would lead to a blackhole link.
Since the BlackholeLink was removed in favor of allowing processors to
drop packets, we should also allow Dispatchers (The Fn that maps a
classificiation to a port number) to drop packets. We do this by
changing the return of a Dispatcher to be an Option type. If it wishes
to drop a packet, it returns a None, rather than Some(portnum).
ClassifyLink then drops the packet in question.

CollinValley committed:
Examples: MSR: Adjust dispatchers to allow drop

The new schema for ClassifyLink allows drops, adjust our dispatcher
implementations to match it. Now our MSR can drop packets that are not
IP type.

CollinValley committed:
Example: DNS-int: Switch to new Dispatcher schema

The new dispatcher requires we return Option<int> to specify port number
to allow classify link to drop packets. Adjust our dispatcher impl accordingly.

moosingin3space committed:
route-rs-afpacket: add links

CollinValley commented on route-rs-afpacket/tests/integration_test.rs:
```diff
@@ -0,0 +1,168 @@
+
+#[test]
+#[ignore]
+fn input_identity() {
```
Why are these tests ignored? If they can't be run by the CI can you leave a small comment about how to run them locally?

CollinValley commented:
Rather than create a whole new crate, let's bring these links into the runtime itself.  They are Links in their own right, and then the runtime can just be dependent on a version of afpacket in the future.

CollinValley committed:
Runtime: runner() now takes a Trait Object

Before we were taking an function object that could produce us a Link
that could be run in the tokio runtime, instead now we take a Trait
Object that implements LinkBuilder. Since it implements LinkBuilder, we
can assume that the object passed in is already initialized, and we then
build the object inside the Tokio runtime context by calling
build_link(). This is much more natural than the previous
implementation, and helps to eliminate some boilerplate.

CollinValley committed:
Runtime: Update run_link to new test_link

This new testing harness allows the user to optionally specify a
timeout. If no timeout is given, test_link will run the link until it
exits. If a timeout is specified, and the link is still running when the
time expires, the link is torn down and the packets that were egressed
from the link are collected and passed to the user. If the link finishes
before the timeout, then it functions as if no timeout were specified.

This is useful for links that do not have an exit condition. The tester
can now let the links run for a set amount of time, and then check that
all the packets were transferred in time.

CollinValley committed:
Minimal Static Router: Update tests to use test_link

MSR needs to use the new test_link function, that features timeouts

CollinValley committed:
PHR: Init example, create main and router

This commit initializes our new Pi Home Router example in the repo,
creating a new cargo crate. It lays out a basic filesystem in a manner
that I think will be useful to the implementors working on the router
internals. The main.rs and router.rs files introduce a limited skeleton
to construct a router. It shows some basic argument parsing, and running
the router with the runner() harness provided by the runtime. The router
object in router.rs does nothing as of yet, and returns an empty Link
object when built.

There are no integration tests created yet for the router, those will
get handled down the line when packet interface intake and exhaust are
implemented.

CollinValley committed:
PHR: Introduce InterfaceCollect and InterfaceDispatch

InterfaceCollect is a new link used to collect packets from all the
input interfaces and join them into a single stream. As it does this, it
also applies an annotation to each packet denoting which interface said
packet originated from. As such, it maps `EthernetFrame` to
`InterfaceAnnotated<EthernetFrame>`.

InterfaceDispatch is a new link that takes in a single stream of
interface annotated packets, and dispatches each packet to the
appropriate outbound interface, removing the interface annotation as it
does so. The outbound_interface annotation is a field of the
InterfaceAnnotated<Packet> structure, and any packet that does not have
a marked outbound interface will be dropped. As such, it maps
`InterfaceAnnotated<EthernetFrame>` to `EthernetFrame`.

As referenced above, this commit also introduces the
InterfaceAnnotated<Packet> structure as a means to demarkate which
interface a packet originated from, and where a packet should be sent
to. This information is needed at many points in the router to determine
the correct course of action for each packet.

Implementation Note: These links could have been implemented in a
generic fashion to faciliate more than the current three
interfaces (Host, LAN, WAN), but that would increase the complexity of
the link. It is not expected that this example will require more
interfaces, but if more are needed, supporting more interfaces shouldn't
be a onerous task.

CollinValley committed:
WIP: PHR: Create RouterIntake and RouterExhaust

Create the Intake and Exhaust Links. These links handle input and output
to the router internally, taking only the AFPacket Recv and Send halfs
as part of their configuration. The top level router object is expected to
create these links, and handle the initialization of AfPacket. If that
exposes too much top level complexity, the AfPacket initialization could
be handled by these links, with some work.

RouterIntake: This link pulls packets from the 3 input interfaces from
AfPacket, tags them by interface, and sorts them into 3 streams by
EtherType. This 3 streams are returned as egressors, and should be
hooked into the ARP, IPv4, and IPv6 router subcomponents. Its type map
is () -> InterfaceAnnotated<EthernetFrame>.

RouterExhaust: This link takes InterfaceAnnotated packets and sends them
to their marked outbound interface. It drops any packets that do not
have a set outbound interface. It supports any number of input streams.
Its type map is InterfaceAnnotated<EthernetFrame> -> ().

EtherType: To enable RouterIntake, we also introduce EtherType and a
related classfier, so we may sort incoming packets by their layer 2
ethertype field.

[WIP: AfPacket Links aren't merged yet, disable compilation of links]
[WIP: Needs tests]

CollinValley committed:
PHR: Rework general file layout

This moves each subcomponent of the router into its own folder. Much
like the link composition, where each larger links are composed of
smaller links, in a similar format. The file layout of the PHR will also
follow a similar pattern, where subdirectories follow a similar layout.
Processors, Classifiers, and Links and other resources are located in
their respective directories.

CollinValley commented on Cargo.toml:
```diff
@@ -8,6 +8,7 @@ members = [
 
     # I/O Crates
     "afpacket",
+    "route-rs-afpacket",
```
Nit: I'd like these links to be moved into the runtime, since all the primitive links reside there, and this is low level enough (Read: actually implements Future/Stream) that I think it has a good home there. Also I would like the runtime crate itself to ship with a complete set of tools, which having I/O links is a requirement of.

I think you should put this whole thing in `route-rs-runtime/links/primitive` and have a cfg-linux flag on the relevant mods.

I am leaving it up to you though, I'll accept wherever you want to stash it in the end, hence Nit.

CollinValley commented on afpacket/src/tokio_sockets.rs:
```diff
@@ -20,11 +35,287 @@ impl AsyncBoundSocket {
 
-    pub async fn send(&mut self, frame: &[u8]) -> io::Result<usize> {
-        self.sock.write(frame).await
+    pub fn split(self) -> (SendHalf, RecvHalf) {
```
Documentation: Something about the purpose of split, what it is used for.  Reference to Tokio maybe.

CollinValley replied:
You don't have to go crazy here, but some documentation is needed for sure. Will help with maintenance in the future as well. :)

CollinValley commented on afpacket/src/tokio_sockets.rs:
```diff
@@ -20,11 +35,287 @@ impl AsyncBoundSocket {
+        (tx, rx)
+    }
+
+    pub fn poll_can_rx(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
```
Docs: When does this return Ready, when does it return Pending.  Add something for each of these poll_* functions. 
Should also specify that these functions are non-blocking.  Just think of all the things you want to know when you work with a library for functions like these.

CollinValley commented on afpacket/src/tokio_sockets.rs:
```diff
@@ -20,11 +35,287 @@ impl AsyncBoundSocket {
+        Poll::Ready(Ok(()))
+    }
+
+    pub fn clear_can_tx(&mut self, cx: &mut Context<'_>) -> io::Result<()> {
```
Maybe explain why one needs to clear?

CollinValley commented on afpacket/src/tokio_sockets.rs:
```diff
@@ -20,11 +35,287 @@ impl AsyncBoundSocket {
+    pub fn poll_recv(
+        &mut self,
+        cx: &mut Context<'_>,
+        frame: &mut [u8],
```
Are there size requirements for frame?  What happens if the data to write would overflow the buffer? Panic? Should be documented.

CollinValley commented on afpacket/src/tokio_sockets.rs:
```diff
@@ -20,11 +35,287 @@ impl AsyncBoundSocket {
+    }
+}
+
+struct RecvHalfRecvFuture<'a> {
```
Let's get some doc comments on what all these structs are for. They won't be visible in the public facing documentation, but useful for maintenance.

CollinValley commented on afpacket/src/tokio_sockets.rs:
```diff
@@ -20,11 +35,287 @@ impl AsyncBoundSocket {
+    guard: Guard<'a>,
+}
+
+impl RecvHalf {
```
Nit: I like Impl blocks near a struct definition if possible.

CollinValley commented on afpacket/tests/afpacket_tokio_tests.rs:
```diff
@@ -65,3 +65,62 @@ fn layer2_loopback() {
+            .unwrap()
+            .unwrap();
+        in_buffer.resize(len, 0);
+        assert_eq!(in_buffer, eth_pkt.data);
```
Asserts need to be outside of the runtime context to cause a test to pass/fail.

CollinValley commented on route-rs-afpacket/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+route-rs-runtime = { path = "../route-rs-runtime" }
+futures = "0.3.4"
+
+[dev-dependencies]
```
Nice touch here with dev-dependencies for test only imports

CollinValley commented on route-rs-afpacket/src/input.rs:
```diff
@@ -0,0 +1,70 @@
+const MTU: usize = 1500;
+
+#[derive(Default)]
+pub struct AfPacketInputLink {
```
Documentation needed for this link.  Especially about how to use it on linux systems, etc.

CollinValley commented on route-rs-afpacket/src/input.rs:
```diff
@@ -0,0 +1,70 @@
+    }
+}
+
+impl link::LinkBuilder<(), Vec<u8>> for AfPacketInputLink {
```
Is it possible for AfPacketInputLink to return half a packet in one recv, then the other half on the next recv? Or does it only return discreet packets?

CollinValley commented on route-rs-afpacket/src/input.rs:
```diff
@@ -0,0 +1,70 @@
+        let me = self.get_mut();
+        let (sz, _) =
+            ready!(me.ingress.poll_recv(cx, &mut me.in_buf)).expect("failed to read packet");
+        me.in_buf.resize(sz, 0);
```
Why do we resize this vector on each received size? Why not just leave it at MTU length always?

CollinValley commented on route-rs-afpacket/src/output.rs:
```diff
@@ -0,0 +1,100 @@
+use std::pin::Pin;
+
+#[derive(Default)]
+pub struct AfPacketOutputLink {
```
Docs for the link.

CollinValley commented on route-rs-afpacket/src/output.rs:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    fn build_link(self) -> link::Link<()> {
+        match (self.in_stream, self.egress) {
```
Nifty way of checking those values are not none, nice.

CollinValley commented on route-rs-afpacket/src/output.rs:
```diff
@@ -0,0 +1,100 @@
+impl Future for StreamToChannel {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
```
Could you doc comment here the cases where it fails to write, how it gets to lock, how it retries, etc?

CollinValley commented on route-rs-afpacket/tests/integration_test.rs:
```diff
@@ -0,0 +1,174 @@
+#[test]
+#[should_panic]
+#[ignore]
+fn output_panics_when_built_without_ingressor() {
```
remove ignore

CollinValley commented on route-rs-afpacket/tests/integration_test.rs:
```diff
@@ -0,0 +1,174 @@
+#[test]
+#[should_panic]
+#[ignore]
+fn output_panics_when_built_with_multiple_ingressors() {
```
remove ignore

CollinValley commented on route-rs-runtime/src/utils/test/harness.rs:
```diff
@@ -2,7 +2,10 @@ use crate::link::{Link, TokioRunnable};
 use crate::utils::test::packet_collectors::ExhaustiveCollector;
```
This whole file needs to be dropped. Was only meant for testing purposes.

CollinValley commented on route-rs-afpacket/tests/integration_test.rs:
```diff
@@ -0,0 +1,174 @@
+            external.send(&ext_pkt_data).await.unwrap();
+        });
+
+        run_link(link).await
```
change to `test_link(link, timeout).await` I'll merge the new test_link() for you by EoW so you can merge this.

CollinValley commented on route-rs-afpacket/tests/integration_test.rs:
```diff
@@ -0,0 +1,174 @@
+            .channel(tx_rtr)
+            .build_link();
+
+        run_link(link).await;
```
similar change to test_link as above.

CollinValley reviewed (commented):
No big changes needed, but doc comments are. Commented what to do with the run_link() function in your integration tests. Overall, the code looks real solid, good work.

CollinValley commented:
This is the commit that changes run_link to test link. https://github.com/route-rs/route-rs/commit/ebdbbaaf9f175a7de7e5b8aeecffaf1ee5d1734d

What I recommend you do, after fixing the above issues, is submit a PR against dev/pi-home-router. That way you have the latest changes while we wait to be ready to merge the runtime changes to master.  Then in a couple weeks we will be back to working off master.

CollinValley committed:
PHR: Rework Intake and Exhaust Links

At our last meeting, other members brought up that we should not code
the IO layers into the business logic of our core router, particularly
since we want to be able to test the router using PCAPs, and swapping
real OS interfaces for PCAP streams shouldn't be a pain. So I am
reworking the intake and exhaust to not do this, and to handle
converting the raw Vec<u8> packet streams to route-rs Packets.

scgruber committed:
ClassifyLink: Create Dispatcher type

Clippy started bugging us about having a declared type for this since
it's pretty complicated.

scgruber committed:
Fix misc clippy and rustfmt issues

This gets the repo state to a place where tests should pass in Travis

scgruber committed:
travis.yml: Test dev/ branches

Fixes the one clippy warning that's still triggering.

CollinValley committed:
PHR: Basic Clippy Fixup

CollinValley committed:
PHR: outbound_interface: Bruh moment cleanup

ddowl committed:
packets: Add ArpFrame wrapper around EthernetFrame

CollinValley committed:
PHR: Router_Intake: Fixup clippy test errors

There were a few clippy messages I missed because they were inside the
tests. This commit resolves those.

scgruber committed:
travis.yml: Fix missing libpcap

The pcap crate requires libpcap, so let's make sure that's installed in
our travis setup.

CollinValley committed:
PHR: Interface: Use more succinct builder syntax

CollinValley committed:
Runtime: Primitive: Use cleaner builder syntax

This syntax is cleaner than before, and doesn't have us redeclaring the
same struct in each function call.

CollinValley committed:
Runtime: Composite: Use better builder syntax

CollinValley committed:
Packets: Allow chaining setters

Adjust the packets library so that setter functions may be chained. This
gives a more ergonomic pseudo-builder style, but doesn't require it,
since references that are uncollected are automatically dropped. I think
this does a good job providing a nice syntactic interface, without
forcing the user into a particular style. It also isn't a breaking
change to the interface.

CollinValley committed:
Unpack!: A Macro to help you remove stuff from links

This is a preliminary cut, for a draft commit
It does two things at once:
For runnables, it takes the vector passed in, your `all_runnables` and
appends the runnables returned from the link to the provided vector. You
may also choose to provide a `_` instead, and the runnables are
discarded.

For egressors, if you provide a single variable, it is declared as your
would expect, and represents the vector of egressors that is returned
from the link. If you provide a one or more variables inside of `[]`
square brackets, it automatically destructures the vector into those the
provided variables.

Examples
unpack!( let (runnables, egressors) = myLink::new().build; );

unpack!( let (runnables, [a, b, c]) = myLink::new().build; );

unpack!( let (_, egressors) = myLink::new().build; );

unpack!( let (_, [a,b,c]) = myLink::new().build; );

CollinValley committed:
Runtime: Primitive: Switch to default unbounded channels

Switch to using unbounded channels inside our primitive channel using
links (Queue, Join, Fork and Classify), if the user does not specify a
channel size. Should they specify a size through the same interface, we
provide them a bounded channel.

This is a good change because the kernel (through the AFPacket
interface) originally allocates the memory for us, we just pass fat
pointers around through the vectors the represent the data in our
packets. Because of this fact, adding and removing things to channels
does not cause an allocation, so it the RAM usage does not really grow
as more and more packets are "added". This would (usually) only happen in the
event that the packet rate exceeds the line rate of our NIC.  The total
number of packets "in-flight" in the router, and hence the total RAM
being used, will be bounded by this ring-buffer exposed to user space.
So users of the runtime shouldn't really have to worry about forcing
various buffer points to be a certain size, the RAM usage off the
program won't grow out of control. This is one of the powers of relying
on Rust, we have good control over the allocations made and can keep the
runtime fast.

Because we are multithreaded, concurrent, and even semi-parallel (multi-cpu), we
should try to let tasks do as much work as possible before they exit,
since this keeps the I-Cache on the CPU hotter. If we set arbitrary
channel size limits, we force the task to switch off before it has
exhausted its input pipe, effectively flushing our cache.

ddowl committed:
packets: Move consts, impl TryFrom, PLS SQUASH

CollinValley committed:
Runtime: Macros: Remove outdated extern crate

Since Rust 2018, we don't have to include extern crate in our libs
anymore, just specifying it in the cargo.toml is enough. Also we can
refer to macros directly in use statements now, rather than with the

CollinValley committed:
Runtime: Unpack!: Add some documentation

ddowl committed:
packets: Chain ArpFrame setters via &mut Self

ddowl committed:
packets: Add failing ArpFrame::TryFrom<EthernetFrame> tests

moosingin3space commented:
Closing this so we can work on it as part of the PHR.
